src/common/timer_wheel.rs

```diff
@@ -237,11 +260,13 @@ impl<K> TimerWheel<K> {
     unsafe fn unlink_timer(&mut self, node: NonNull<DeqNode<TimerNode<K>>>) {
         let p = node.as_ref();
         if let TimerNode::Entry { level, index, .. } = &p.element {
-            let level = level.load(Ordering::Acquire);
-            let index = index.load(Ordering::Acquire);
-            if level != u8::MAX && index != u8::MAX {
-                self.wheels[level as usize][index as usize].unlink(node);
+            let lev = level.load(Ordering::Acquire);
+            let idx = index.load(Ordering::Acquire);
+            if lev != u8::MAX && idx != u8::MAX {
+                self.wheels[lev as usize][idx as usize].unlink(node);
             }
+            level.store(u8::MAX, Ordering::Release);
+            index.store(u8::MAX, Ordering::Release);
```

Comment:
I’m absolutely no expert on atomics, etc, but I’m curious here:
- Why is this two separate `AtomicU8` values, instead of a single `AtomicU16` with two masked values?
- As these are not cacheline-aligned, wouldn’t these two separate values clash directly? Probably not a big deal, but as multiple threads can probably interleave here, two load + two store operations per thread have a higher likelyhood to slow things down, no?
- I believe there is still a race condition between the `load` and, `unlink` and `store`. Two threads can first load the values and execute the `unlink`, before either thread sets that to `u8::MAX`
- Wouldn’t it be better to use an atomic exchange to change this to `u8::MAX` (or `u16::MAX` if you merge the two flags) before checking the flag, so that any competing thread will never execute `unlink` twice?

Reply:
Thank you for reviewing!

As for using two separate `AtomicU8` values, there is not much reason; I just wanted two integers that are `Send` + `Sync`. As you said, using a single `AtomicU16` should be more efficient (and atomic) so I will change them so.

As for the race condition, we do not have to be bothered in `TimerWheel` code because it is guaranteed that no more than one thread will call a method of single `TimerWheel` instance at the same time. About half the code in `base_cache` module are executed by the `sync` method of `Inner` type. And `sync` method is guarded by a cache-wide lock `self.deques.lock()`. This is by design; `sync` applies cache reads/writes logs in batch to reduce the lock contentions.

Other data structures including `Deque`s and `FrequencySketch` are guarded by the same cache-wide lock.

Maybe I should add source code comments and/or divide the module into smaller to make this clear.

Reply:
It turned out I did not need interior mutability for them 😅 because the `TimerWheel`, the owner of `TimerNode`s, is guarded by a mutex. I replaced these two `AtomicU8`s with a single `Option<(u8, u8)>`.

I am running some load tests now. I will check-in the change once I get some confidence.

Reply:
> I replaced these two `AtomicU8`s with a single `Option<(u8, u8)>`.

OK. I checked-in the change: [`ac4d32d`#diff-L86-R86](https://github.com/moka-rs/moka/pull/248/commits/ac4d32de4ef1deb925c50c227cbccec957effcf4#diff-a9ca7964752b28bb55848a11e88855234c65c66430b6fc82f2dacaf265c55fc8L86-R86)